### PR TITLE
Adopt AssertJ duration assertions

### PR DESCRIPTION
**Pull request by AlekSimpson**
## What's changed?
This PR introduces a recipe that simplifies some AssertJ duration assertions with dedicated asserts and also simplifies some `Duration`s to easier to read forms.
Ex:
```
- .hasHours(48)
+ .hasDays(2)
```

## What's your motivation?
[#400](https://github.com/openrewrite/rewrite-testing-frameworks/issues/400)

## Anyone you would like to review specifically?
@timtebeek 

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've added the license header to any new files through `./gradlew licenseFormat`
- [x] I've used the IntelliJ IDEA auto-formatter on affected files
- [ ] I've updated the documentation (if applicable)

### CODE REVIEW

**Comment by timtebeek** [comment]
We might also want to cover these methods: [isNegative](https://www.javadoc.io/static/org.assertj/assertj-core/3.24.2/org/assertj/core/api/AbstractDurationAssert.html#isNegative()), [isPositive](https://www.javadoc.io/static/org.assertj/assertj-core/3.24.2/org/assertj/core/api/AbstractDurationAssert.html#isPositive()), [isZero](https://www.javadoc.io/static/org.assertj/assertj-core/3.24.2/org/assertj/core/api/AbstractDurationAssert.html#isZero()). 
Other than that a good start, after fixing the compilation errors in 99c123c.

You can [reuse this (now) public method to simplify the expressions to a constant value](https://github.com/openrewrite/rewrite-static-analysis/blob/ba080f5e7fb3698cc196e6199fbed22822e720f2/src/main/java/org/openrewrite/staticanalysis/SimplifyDurationCreationUnits.java#L116):.

File: src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java [hunk 1]
```diff
@@ -27,4 +57,53 @@ public String getDisplayName() {
     public String getDescription() {
         return "Adopt AssertJ DurationAssert assertions for more expressive messages.";
     }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new AdoptAssertJDurationAssertionsVisitor();
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java:63]: Can we add a precondtion here on using either `org.assertj.core.api.DurationAssert has*(..)` or `org.assertj.core.api.AbstractLongAssert isEqualTo(..)`? That could help prevent the `stream().anyMatch` from being evaluated for each method invocation.

Comment (timtebeek) [reply src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java:63]: Even a matcher on `org.assertj.core.api.Assertions assertThat(..)` could limit execution quite a bit already, but more specific is better I think

File: src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    );
+
+    @Override
+    public String getDisplayName() {
+        return "Adopt AssertJ Duration assertions";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Adopt AssertJ DurationAssert assertions for more expressive messages.";
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java:61]:
```suggestion
        return "Adopt AssertJ `DurationAssert` assertions for more expressive messages.";
```

File: src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return Preconditions.check(Preconditions.or(
+                new UsesMethod<>("org.assertj.core.api.AbstractDurationAssert has*(int)"),
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java:67]:
```suggestion
                new UsesMethod<>("org.assertj.core.api.AbstractDurationAssert has*(int)", true),
```

File: src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+                new UsesMethod<>("org.assertj.core.api.AbstractLongAssert isEqualTo(..)")
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java:68]:
```suggestion
                new UsesMethod<>("org.assertj.core.api.AbstractLongAssert isEqualTo(..)", true)
```

File: src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+            Long argValue = SimplifyDurationCreationUnits.getConstantIntegralValue(arg); // note: guess my machine hasn't updated to the new commit yet, method should be public
+            if (argValue % timeLength == 0) {
+                // convert divided value to OpenRewrite LST type
+                int rawValue = (int)(argValue / timeLength);
+                J.Literal newArg = rawValueToExpression(rawValue);
+                // get new method name
+
+                // update method invocation with new name and arg
+                Expression methodSelect = m.getSelect();
+                return JavaTemplate.builder("#{any()}.#{any()}(#{int})")
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java:132]: Just as a hint:
```suggestion
                return JavaTemplate.builder("#{any()}.#{any()}(#{any(int)})")
```

Comment (AlekSimpson) [reply src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java:132]: ah whoops, I don't know why that slipped my mind. Thanks for catching it.

File: src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    static final MethodMatcher ASSERT_THAT_MATCHER = new MethodMatcher("org.assertj.core.api.Assertions assertThat(..)");
+    static final MethodMatcher GET_NANO_MATCHER = new MethodMatcher("java.time.Duration getNano()");
+    static final MethodMatcher GET_SECONDS_MATCHER = new MethodMatcher("java.time.Duration getSeconds()");
+    static final MethodMatcher ISEQUALTO_MATCHER = new MethodMatcher("org.assertj.core.api.AbstractLongAssert isEqualTo(..)");
+    static final Map<String, String> methodMap = new HashMap<String, String>() {{
+        put("getSeconds", "hasSeconds");
+        put("getNano", "hasNanos");
+    }};
+    static List<MethodMatcher> timeUnitMatchers = Arrays.asList(
+            new MethodMatcher("org.assertj.core.api.AbstractDurationAssert hasMillis(..)", true),
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java:46]: Please note that the following five lines should probably be changed in the same way. I just added the suggestion to the first one.

File: src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+            int methodArg = (int)unitInfo.get(1);
+
+            if (!(m.getSimpleName().equals(methodName))) {
+                // convert divided value to OpenRewrite LST type
+                J.Literal newArg = rawValueToExpression(methodArg);
+                // update method invocation with new name and arg
+                Expression methodSelect = m.getSelect();
+                return JavaTemplate.builder("#{any()}.#{}(#{any(int)})")
+                        .contextSensitive()
+                        .javaParser(JavaParser.fromJavaVersion().classpathFromResources(ctx, "junit-jupiter-api-5.9", "assertj-core-3.24"))
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java:131]: When would we need ` "junit-jupiter-api-5.9"` here and above?

File: src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    @SuppressWarnings("DataFlowIssue")
+    private static class AdoptAssertJDurationAssertionsVisitor extends JavaIsoVisitor<ExecutionContext> {
+        @Override
+        public J.MethodInvocation visitMethodInvocation(J.MethodInvocation mi, ExecutionContext ctx) {
+            J.MethodInvocation m = super.visitMethodInvocation(mi, ctx);
+            if (timeUnitMatchers.stream().anyMatch(matcher -> matcher.matches(mi))) {
+                return simplifyTimeUnits(mi, ctx);
+            }else if (ISEQUALTO_INT_MATCHER.matches(mi) || ISEQUALTO_LONG_MATCHER.matches(mi)) {
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java:84]: Formatter not yet applied here.

File: src/test/java/org/openrewrite/java/testing/assertj/AdoptAssertJDuractionAssertionsTest.java [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+                  void testMethod(Duration time) {
+                      assertThat(time).hasMinutes(5);
+                  }
+              }
+              """
+          )
+        );
+    }
+
+    //millis to hours
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/java/testing/assertj/AdoptAssertJDuractionAssertionsTest.java:202]: I don't think these comments add much; can we remove these?

Comment (AlekSimpson) [reply src/test/java/org/openrewrite/java/testing/assertj/AdoptAssertJDuractionAssertionsTest.java:202]: oh yes, I just had them there to help me when I was writing them all out

File: src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java (hunk 8 above)
Comment (AlekSimpson) [reply src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java:84]: Not quite sure I am clear on this, do you mean that `maybeAutoFormat` should be used here?

Comment (timtebeek) [reply src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java:84]: I mean you've not run IntelliJ IDEA `Cmd + Alt + L`.

**Comment by AlekSimpson** [comment]
@timtebeek Made all the revisions here except for the format one that I am unclear on. I also wanted to ask if you had specific replacements in mind for isPositive, isNegative and isZero. I assume you mean replacements like:

```
-.isEqualTo(0)
+.isZero()
```

**Comment by timtebeek** [comment]
> I also wanted to ask if you had specific replacements in mind for isPositive, isNegative and isZero. I assume you mean replacements like:
> 
> ```
> -.isEqualTo(0)
> +.isZero()
> ```

Indeed such replacements are what I had in mind, as we go from comparing numbers to the [isNegative](https://www.javadoc.io/static/org.assertj/assertj-core/3.24.2/org/assertj/core/api/AbstractDurationAssert.html#isNegative()), [isPositive](https://www.javadoc.io/static/org.assertj/assertj-core/3.24.2/org/assertj/core/api/AbstractDurationAssert.html#isPositive()) & [isZero](https://www.javadoc.io/static/org.assertj/assertj-core/3.24.2/org/assertj/core/api/AbstractDurationAssert.html#isZero()) methods available on `AbstractDurationAssert`

**Comment by timtebeek** [comment]
H @AlekSimpson ; I've added a few commits just now as I've gone over the implementation; I've also added some tests that highlight existing issues. Could you fix those before we merge?

**Comment by AlekSimpson** [comment]
@timtebeek I am currently updating the code to handle the `.as()` test case. I've made it so that if the `as()` method is found then it gets the argument and saves it to add to the template parameters and then the template that is passed into the `JavaTemplate.builder()` is updated to retain the `as()` method. I keep getting the error:

**Comment by knutwannheden** [comment]
There is a `%s` in that error message. I assume something didn't get substituted before you compiled the template.

**Comment by AlekSimpson** [comment]
> There is a `%s` in that error message. I assume something didn't get substituted before you compiled the template.

Oh yep, that's definitely the issue

**Comment by AlekSimpson** [comment]
@timtebeek The latest commit fixes two of the three tests you added. The last one that is not fixed is the retain whitespace test case. Do I need to use something like `maybeAutoformat` to fix this test? I considered trying to somehow manually add the whitespace in but that didn't seem like a very flexible solution to me.

File: src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+                return Arrays.asList(name, argValue);
+            }
+        }
+
+        private J.MethodInvocation applyTemplate(ExecutionContext ctx, J.MethodInvocation m, String template, Object... parameters) {
+            return JavaTemplate.builder(template)
+                    .contextSensitive()
+                    .javaParser(JavaParser.fromJavaVersion().classpathFromResources(ctx, "assertj-core-3.24"))
+                    .build()
+                    .apply(getCursor(), m.getCoordinates().replace(), parameters);
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java:180]: @AlekSimpson What you want to do is to retain the whitespace here. Assuming you know that the result has a non-null `select` you can achieve that by doing something like this:
```java
            J.MethodInvocation invocation = JavaTemplate.builder(template)
                    .contextSensitive()
                    .javaParser(JavaParser.fromJavaVersion().classpathFromResources(ctx, "assertj-core-3.24"))
                    .build()
                    .apply(getCursor(), m.getCoordinates().replace(), parameters);
            return invocation.getPadding()
                    .withSelect(
                            invocation.getPadding().getSelect()
                                    .withAfter(m.getPadding().getSelect().getAfter())
                    );
```
But it would probably be best if you guard this with null checks for the selects on both elements.

Comment (AlekSimpson) [reply src/main/java/org/openrewrite/java/testing/assertj/AdoptAssertJDurationAssertions.java:180]: This works! Thanks for pointing it out, I'll be sure to add those null checks in.

**Review by timtebeek — APPROVED** [review]

**Comment by Bananeweizen** [comment]
I would like to suggest moving the isZero, isPositive, isNegative out of this recipe. Here they apply just for duration, right? But they are valid for any assertion of numeric type, and would have far greater impact then.

**Comment by knutwannheden** [comment]
> I would like to suggest moving the isZero, isPositive, isNegative out of this recipe. Here they apply just for duration, right? But they are valid for any assertion of numeric type, and would have far greater impact then.

Sounds good! Do you want to provide a PR for this? 😄
